<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
        <title>RNA-seq Ref-align Analysis (1) - Coffe Chats</title>
        <meta name="Description" content=""><meta property="og:title" content="RNA-seq Ref-align Analysis (1)" />
<meta property="og:description" content="A routine analysis of RNA-seq." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://zkhuang2012.github.io/mrna_ref1/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2020-05-08T21:30:50+08:00" />
<meta property="article:modified_time" content="2020-05-08T21:30:50+08:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="RNA-seq Ref-align Analysis (1)"/>
<meta name="twitter:description" content="A routine analysis of RNA-seq."/>
<meta name="application-name" content="Huang ZK&#39;s Blog">
<meta name="apple-mobile-web-app-title" content="Huang ZK&#39;s Blog"><meta name="theme-color" content="#ffffff"><meta name="msapplication-TileColor" content="#da532c"><link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
        <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png"><link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png"><link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5"><link rel="manifest" href="/site.webmanifest"><link rel="canonical" href="https://zkhuang2012.github.io/mrna_ref1/" /><link rel="prev" href="https://zkhuang2012.github.io/about/" /><link rel="stylesheet" href="/lib/normalize/normalize.min.css"><link rel="stylesheet" href="/css/style.en.min.css"><link rel="stylesheet" href="/lib/fontawesome-free/all.min.css"><link rel="stylesheet" href="/lib/animate/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "RNA-seq Ref-align Analysis (1)",
        "inLanguage": "en",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "https:\/\/zkhuang2012.github.io\/mrna_ref1\/"
        },"image": {
                "@type": "ImageObject",
                "url": "https:\/\/zkhuang2012.github.io\/cover.png",
                "width":  800 ,
                "height":  600 
            },"genre": "posts","keywords": "RNA-seq, HISAT2, clusterProfiler","wordcount":  909 ,
        "url": "https:\/\/zkhuang2012.github.io\/mrna_ref1\/","datePublished": "2020-05-08T21:30:50+08:00","dateModified": "2020-05-08T21:30:50+08:00","license": "This work is licensed under a Creative Commons Attribution-NonCommercial 4.0 International License.","publisher": {
                "@type": "Organization",
                "name": "zkhuang's Blog",
                "logo": {
                "@type": "ImageObject",
                "url": "https:\/\/zkhuang2012.github.io\/logo.png",
                "width":  127 ,
                "height":  40 
                }
            },"author": {
                "@type": "Person",
                "name": "zkhuang"
            },"description": ""
    }
    </script></head>
    <body><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('auto' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : 'auto' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="/" title="Coffe Chats"><span class="header-title-pre"><i class='far fa-coffee fa'></i></span><span id="id-1" class="typeit"></span></a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="/posts/"> Posts </a><a class="menu-item" href="/tags/"> Tags </a><a class="menu-item" href="/categories/"> Categories </a><a class="menu-item" href="/about/"> About </a><a class="menu-item" href="https://github.com/zkhuang2012" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw'></i>  </a><span class="menu-item delimiter"></span><a href="javascript:void(0);" class="menu-item language" title="Select Language">English<i class="fas fa-chevron-right fa-fw"></i>
                        <select class="language-select" id="language-select-desktop" onchange="location = this.value;"><option value="/mrna_ref1/" selected>English</option><option value="/zh-cn/mrna_ref1/">简体中文</option></select>
                    </a><span class="menu-item search" id="search-desktop">
                        <input type="text" placeholder="Search titles or contents..." id="search-input-desktop">
                        <a href="javascript:void(0);" class="search-button search-toggle" id="search-toggle-desktop" title="Search">
                            <i class="fas fa-search fa-fw"></i>
                        </a>
                        <a href="javascript:void(0);" class="search-button search-clear" id="search-clear-desktop" title="Clear">
                            <i class="fas fa-times-circle fa-fw"></i>
                        </a>
                        <span class="search-button search-loading" id="search-loading-desktop">
                            <i class="fas fa-spinner fa-fw fa-spin"></i>
                        </span>
                    </span><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                    <i class="fas fa-adjust fa-fw"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="/" title="Coffe Chats"><span class="header-title-pre"><i class='far fa-coffee fa'></i></span><span id="id-2" class="typeit"></span></a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><div class="search-wrapper">
                    <div class="search mobile" id="search-mobile">
                        <input type="text" placeholder="Search titles or contents..." id="search-input-mobile">
                        <a href="javascript:void(0);" class="search-button search-toggle" id="search-toggle-mobile" title="Search">
                            <i class="fas fa-search fa-fw"></i>
                        </a>
                        <a href="javascript:void(0);" class="search-button search-clear" id="search-clear-mobile" title="Clear">
                            <i class="fas fa-times-circle fa-fw"></i>
                        </a>
                        <span class="search-button search-loading" id="search-loading-mobile">
                            <i class="fas fa-spinner fa-fw fa-spin"></i>
                        </span>
                    </div>
                    <a href="javascript:void(0);" class="search-cancel" id="search-cancel-mobile">
                        Cancel
                    </a>
                </div><a class="menu-item" href="/posts/" title="">Posts</a><a class="menu-item" href="/tags/" title="">Tags</a><a class="menu-item" href="/categories/" title="">Categories</a><a class="menu-item" href="/about/" title="">About</a><a class="menu-item" href="https://github.com/zkhuang2012" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw'></i></a><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                <i class="fas fa-adjust fa-fw"></i>
            </a><a href="javascript:void(0);" class="menu-item" title="Select Language">English<i class="fas fa-chevron-right fa-fw"></i>
                    <select class="language-select" onchange="location = this.value;"><option value="/mrna_ref1/" selected>English</option><option value="/zh-cn/mrna_ref1/">简体中文</option></select>
                </a></div>
    </div>
</header>
<div class="search-dropdown desktop">
    <div id="search-dropdown-desktop"></div>
</div>
<div class="search-dropdown mobile">
    <div id="search-dropdown-mobile"></div>
</div>
<main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">Contents</h2>
            <div class="toc-content" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">RNA-seq Ref-align Analysis (1)</h1><div class="post-meta">
            <div class="post-meta-line"><span class="post-author"><a href="/" title="Author" rel=" author" class="author"><i class="fas fa-user-circle fa-fw"></i>zkhuang</a></span>&nbsp;
                    <span class="post-category">included in<a href="/categories/rna-seq/">
                                <i class="far fa-folder fa-fw"></i>RNA-seq
                            </a></span></div>
            <div class="post-meta-line"><i class="far fa-calendar-alt fa-fw"></i><time datetime=2020-05-08>2020-05-08</time>&nbsp;
                <i class="fas fa-pencil-alt fa-fw"></i>about 909 words&nbsp;
                <i class="far fa-clock fa-fw"></i>5 min&nbsp;</div>
        </div><div class="details toc" id="toc-static">
                <div class="details-summary toc-title">
                    <span>Contents</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a></li>
    <li><a href="#1st-prepariation">1st, Prepariation</a></li>
    <li><a href="#2nd-alignment">2nd, Alignment</a></li>
    <li><a href="#3rd-estimate-transcript-abundance">3rd, Estimate transcript abundance</a></li>
    <li><a href="#4th-differential-expression-analysis">4th, Differential expression analysis</a></li>
    <li><a href="#5th-funtional-enrichment-analysis">5th, Funtional enrichment analysis</a></li>
    <li><a href="#reference">Reference</a></li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><h2 id="introduction">Introduction</h2>
<p>This RNA-seq pipeline was built with HISAT2 + Stringtie + DESeq2 + clusterProfiler.</p>
<h2 id="1st-prepariation">1st, Prepariation</h2>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="c1"># Directory Deployment</span>
mkdir -p <span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/project/mRNA/<span class="o">{</span>bin,data/<span class="o">{</span>genes,genome,index,samples<span class="o">}</span>,output/<span class="o">{</span>ballgown,hisat2,tmp<span class="o">}}</span>


<span class="c1"># Data link</span>
<span class="nb">cd</span> <span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/../Data_benthic/practice/mRNA
ln -s <span class="si">${</span><span class="nv">PWD</span><span class="si">}</span>/Hdh.fa <span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/project/mRNA/data/genome/
ln -s <span class="si">${</span><span class="nv">PWD</span><span class="si">}</span>/Hdh.gff3 <span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/project/mRNA/data/genes/

<span class="k">for</span> i in <span class="sb">`</span>find <span class="si">${</span><span class="nv">PWD</span><span class="si">}</span> -type f -name <span class="s2">&#34;*.fq.gz&#34;</span><span class="sb">`</span><span class="p">;</span>
<span class="k">do</span>
	ln -s <span class="si">${</span><span class="nv">i</span><span class="si">}</span> <span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/project/mRNA/data/samples/<span class="p">;</span>
<span class="k">done</span>

</code></pre></td></tr></table>
</div>
</div><h2 id="2nd-alignment">2nd, Alignment</h2>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="nb">cd</span> <span class="si">${</span><span class="nv">HOME</span><span class="si">}</span>/project/mRNA/bin
</code></pre></td></tr></table>
</div>
</div><p>Copy the codes below to <strong>02alignment.pbs</strong></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span><span class="lnt">42
</span><span class="lnt">43
</span><span class="lnt">44
</span><span class="lnt">45
</span><span class="lnt">46
</span><span class="lnt">47
</span><span class="lnt">48
</span><span class="lnt">49
</span><span class="lnt">50
</span><span class="lnt">51
</span><span class="lnt">52
</span><span class="lnt">53
</span><span class="lnt">54
</span><span class="lnt">55
</span><span class="lnt">56
</span><span class="lnt">57
</span><span class="lnt">58
</span><span class="lnt">59
</span><span class="lnt">60
</span><span class="lnt">61
</span><span class="lnt">62
</span><span class="lnt">63
</span><span class="lnt">64
</span><span class="lnt">65
</span><span class="lnt">66
</span><span class="lnt">67
</span><span class="lnt">68
</span><span class="lnt">69
</span><span class="lnt">70
</span><span class="lnt">71
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="cp">#!/bin/bash -x
</span><span class="cp"></span><span class="c1">#PBS -N ref_mRNA_02alignment</span>
<span class="c1">#PBS -o ref_mRNA_02alignment.out</span>
<span class="c1">#PBS -e ref_mRNA_02alignment.err</span>
<span class="c1">#PBS -q high</span>
<span class="c1">#PBS -j oe</span>
<span class="c1">#PBS -l nodes=1:ppn=28</span>

<span class="nv">NUMCPUS</span><span class="o">=</span><span class="sb">`</span>wc -l &lt;<span class="nv">$PBS_NODEFILE</span><span class="sb">`</span>
<span class="nv">THREADS</span><span class="o">=</span><span class="sb">`</span>expr <span class="si">${</span><span class="nv">NUMCPUS</span><span class="si">}</span> <span class="se">\*</span> 2<span class="sb">`</span>
<span class="nv">BASEDIR</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">HOME</span><span class="si">}</span><span class="s2">/project/mRNA&#34;</span>
<span class="nv">FASTQLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/samples&#34;</span>
<span class="nv">GENOMEIDX</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/index/Hdh_tran&#34;</span>
<span class="nv">GTFFILE</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/genes/Hdh.gtf&#34;</span>
<span class="nv">GENOMEFA</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/genome/Hdh.fa&#34;</span>
<span class="nv">WRKDIR</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/output&#34;</span>
<span class="nv">ALIGNLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span><span class="s2">/hisat2&#34;</span>
<span class="nv">BALLGOWNLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span><span class="s2">/ballgown&#34;</span>
<span class="nv">TEMPLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span><span class="s2">/tmp&#34;</span>

<span class="c1"># transfer gff to gtf</span>
<span class="nb">cd</span> <span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span>/data/genes/
gffread Hdh.gff -T -o <span class="si">${</span><span class="nv">GTFFILE</span><span class="si">}</span>

<span class="c1"># Extract splice sites and exon info from gtf using the script under hisat_install_dir</span>
<span class="nb">cd</span> <span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span>/data/index
python2 /public/tools/rna_seq/hisat2-2.1.0/extract_splice_sites.py <span class="si">${</span><span class="nv">GTFFILE</span><span class="si">}</span> &gt; Hdh.ss
python2 /public/tools/rna_seq/hisat2-2.1.0/extract_exons.py <span class="si">${</span><span class="nv">GTFFILE</span><span class="si">}</span> &gt; Hdh.exon

<span class="c1"># Build genome index for hisat2</span>
hisat2-build -p <span class="m">50</span> --ss Hdh.ss --exon Hdh.exon <span class="si">${</span><span class="nv">GENOMEFA</span><span class="si">}</span> Hdh_tran

<span class="c1">## sample pair-end fastq must be named in this format “_1.*”&#34;_2.*&#34;. eg, &#34;gill01_1.fastq.gz&#34;,&#34;gill01_2.fastq.gz&#34;</span>
<span class="nv">reads1</span><span class="o">=(</span><span class="si">${</span><span class="nv">FASTQLOC</span><span class="si">}</span>/*_1.*<span class="o">)</span>
<span class="nv">reads1</span><span class="o">=(</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">reads1</span><span class="p">[@]##*/</span><span class="si">}</span><span class="s2">&#34;</span><span class="o">)</span>
<span class="nv">reads2</span><span class="o">=(</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">reads1</span><span class="p">[@]/_1./_2.</span><span class="si">}</span><span class="s2">&#34;</span><span class="o">)</span>


<span class="nb">cd</span> <span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span>

<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;#&gt; START: HISAT2 Alignment&#34;</span>
<span class="k">for</span> <span class="o">((</span><span class="nv">i</span><span class="o">=</span>0<span class="p">;</span> i&lt;<span class="o">=</span><span class="si">${#</span><span class="nv">reads1</span><span class="p">[@]</span><span class="si">}</span>-1<span class="p">;</span> i++ <span class="o">))</span><span class="p">;</span> <span class="k">do</span>
	<span class="nv">sample</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">reads1</span><span class="p">[</span><span class="nv">$i</span><span class="p">]%%.*</span><span class="si">}</span><span class="s2">&#34;</span>
	<span class="nv">sample</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">sample</span><span class="p">%_*</span><span class="si">}</span><span class="s2">&#34;</span>

	<span class="nv">stime</span><span class="o">=</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span>
	<span class="nb">echo</span> <span class="s2">&#34;[</span><span class="nv">$stime</span><span class="s2">] Processing sample: </span><span class="nv">$sample</span><span class="s2">&#34;</span>
	<span class="nb">echo</span> <span class="o">[</span><span class="nv">$stime</span><span class="o">]</span> <span class="s2">&#34;   * Alignment of reads to genome (HISAT2)&#34;</span>

	hisat2 -p <span class="nv">$NUMCPUS</span> <span class="se">\
</span><span class="se"></span>		--dta <span class="se">\
</span><span class="se"></span>		-x <span class="si">${</span><span class="nv">GENOMEIDX</span><span class="si">}</span> <span class="se">\
</span><span class="se"></span>		-1 <span class="si">${</span><span class="nv">FASTQLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">reads1</span><span class="p">[</span><span class="nv">$i</span><span class="p">]</span><span class="si">}</span> <span class="se">\
</span><span class="se"></span>		-2 <span class="si">${</span><span class="nv">FASTQLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">reads2</span><span class="p">[</span><span class="nv">$i</span><span class="p">]</span><span class="si">}</span> <span class="se">\
</span><span class="se"></span>		-S <span class="si">${</span><span class="nv">TEMPLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.sam 2&gt;<span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.alnstats

<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;   * Alignments conversion (SAMTools)&#34;</span>
	samtools view -S -b <span class="si">${</span><span class="nv">TEMPLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.sam <span class="p">|</span> <span class="se">\
</span><span class="se"></span>		samtools sort -@ <span class="nv">$NUMCPUS</span> -o <span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.bam -

	rm <span class="si">${</span><span class="nv">TEMPLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.sam

<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;   * Assemble transcripts (StringTie)&#34;</span>
	stringtie -p <span class="nv">$NUMCPUS</span> <span class="se">\
</span><span class="se"></span>		-G <span class="si">${</span><span class="nv">GTFFILE</span><span class="si">}</span> <span class="se">\
</span><span class="se"></span>		-o <span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.gtf <span class="se">\
</span><span class="se"></span>		-l <span class="si">${</span><span class="nv">sample</span><span class="si">}</span> <span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.bam

<span class="k">done</span>

<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;#&gt; Step Two : Alignment Finished!&#34;</span>
</code></pre></td></tr></table>
</div>
</div><p>Then run via PBS:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell">qsub 02alignment.pbs

</code></pre></td></tr></table>
</div>
</div><h2 id="3rd-estimate-transcript-abundance">3rd, Estimate transcript abundance</h2>
<p>Copy the codes below to <strong>03estimation.pbs</strong></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span><span class="lnt">42
</span><span class="lnt">43
</span><span class="lnt">44
</span><span class="lnt">45
</span><span class="lnt">46
</span><span class="lnt">47
</span><span class="lnt">48
</span><span class="lnt">49
</span><span class="lnt">50
</span><span class="lnt">51
</span><span class="lnt">52
</span><span class="lnt">53
</span><span class="lnt">54
</span><span class="lnt">55
</span><span class="lnt">56
</span><span class="lnt">57
</span><span class="lnt">58
</span><span class="lnt">59
</span><span class="lnt">60
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="cp">#!/bin/bash -x
</span><span class="cp"></span><span class="c1">#PBS -N ref_mRNA_03estimation</span>
<span class="c1">#PBS -o ref_mRNA_03estimation.out</span>
<span class="c1">#PBS -e ref_mRNA_03estimation.err</span>
<span class="c1">#PBS -q high</span>
<span class="c1">#PBS -j oe</span>
<span class="c1">#PBS -l nodes=1:ppn=28</span>

<span class="nv">NUMCPUS</span><span class="o">=</span><span class="sb">`</span>wc -l &lt;<span class="nv">$PBS_NODEFILE</span><span class="sb">`</span>
<span class="nv">THREADS</span><span class="o">=</span><span class="sb">`</span>expr <span class="si">${</span><span class="nv">NUMCPUS</span><span class="si">}</span> <span class="se">\*</span> 2<span class="sb">`</span>
<span class="nv">BASEDIR</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">HOME</span><span class="si">}</span><span class="s2">/project/mRNA&#34;</span>
<span class="nv">FASTQLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/samples&#34;</span>
<span class="nv">GENOMEIDX</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/index/Hdh_tran&#34;</span>
<span class="nv">GTFFILE</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/genes/Hdh.gtf&#34;</span>
<span class="nv">GENOMEFA</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/data/genome/Hdh.fa&#34;</span>
<span class="nv">WRKDIR</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">BASEDIR</span><span class="si">}</span><span class="s2">/output&#34;</span>
<span class="nv">ALIGNLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span><span class="s2">/hisat2&#34;</span>
<span class="nv">BALLGOWNLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span><span class="s2">/ballgown&#34;</span>
<span class="nv">TEMPLOC</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span><span class="s2">/tmp&#34;</span>

<span class="nv">reads1</span><span class="o">=(</span><span class="si">${</span><span class="nv">FASTQLOC</span><span class="si">}</span>/*_1.*<span class="o">)</span>
<span class="nv">reads1</span><span class="o">=(</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">reads1</span><span class="p">[@]##*/</span><span class="si">}</span><span class="s2">&#34;</span><span class="o">)</span>
<span class="nv">reads2</span><span class="o">=(</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">reads1</span><span class="p">[@]/_1./_2.</span><span class="si">}</span><span class="s2">&#34;</span><span class="o">)</span>

<span class="nb">cd</span> <span class="si">${</span><span class="nv">WRKDIR</span><span class="si">}</span>

<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;#&gt; START: STRINGTIE Estimation&#34;</span>

<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;#&gt; Merge all transcripts (StringTie)&#34;</span>
ls -1 <span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/*.gtf &gt; <span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/mergelist.txt

stringtie --merge <span class="se">\
</span><span class="se"></span>	-p <span class="nv">$NUMCPUS</span> <span class="se">\
</span><span class="se"></span>	-G  <span class="si">${</span><span class="nv">GTFFILE</span><span class="si">}</span> <span class="se">\
</span><span class="se"></span>	-o <span class="si">${</span><span class="nv">BALLGOWNLOC</span><span class="si">}</span>/stringtie_merged.gtf <span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/mergelist.txt

<span class="c1">## Estimate transcript abundance</span>
<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;#&gt; Estimate abundance for each sample (StringTie)&#34;</span>
<span class="k">for</span> <span class="o">((</span><span class="nv">i</span><span class="o">=</span>0<span class="p">;</span> i&lt;<span class="o">=</span><span class="si">${#</span><span class="nv">reads1</span><span class="p">[@]</span><span class="si">}</span>-1<span class="p">;</span> i++ <span class="o">))</span><span class="p">;</span> <span class="k">do</span>

	<span class="nv">sample</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">reads1</span><span class="p">[</span><span class="nv">$i</span><span class="p">]%%.*</span><span class="si">}</span><span class="s2">&#34;</span>
	<span class="nv">sample</span><span class="o">=</span><span class="s2">&#34;</span><span class="si">${</span><span class="nv">sample</span><span class="p">%_*</span><span class="si">}</span><span class="s2">&#34;</span>

	<span class="k">if</span> <span class="o">[</span> ! -d <span class="si">${</span><span class="nv">BALLGOWNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span> <span class="o">]</span><span class="p">;</span> <span class="k">then</span>
		mkdir -p <span class="si">${</span><span class="nv">BALLGOWNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>
	<span class="k">fi</span>

<span class="c1"># Estimate transcript abundance depend on the referenge genome with ignoring novel transcripts</span>
	stringtie -e -B <span class="se">\
</span><span class="se"></span>		-p <span class="nv">$NUMCPUS</span> <span class="se">\
</span><span class="se"></span>		-G <span class="si">${</span><span class="nv">GTFFILE</span><span class="si">}</span> <span class="se">\
</span><span class="se"></span>		-o <span class="si">${</span><span class="nv">BALLGOWNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.gtf <span class="si">${</span><span class="nv">ALIGNLOC</span><span class="si">}</span>/<span class="si">${</span><span class="nv">sample</span><span class="si">}</span>.bam

<span class="k">done</span>

<span class="c1"># Generate transcript_count_matrix.csv and gene_count_matrix.csv using the script under stringtie_install_dir</span>

python2 /public/tools/rna_seq/stringtie-1.3.4/prepDE.py -l <span class="m">150</span>

<span class="nb">echo</span> <span class="o">[</span><span class="sb">`</span>date +<span class="s2">&#34;%Y-%m-%d %H:%M:%S&#34;</span><span class="sb">`</span><span class="o">]</span> <span class="s2">&#34;#&gt; Step Three: Estimation Finished!&#34;</span>
</code></pre></td></tr></table>
</div>
</div><p>Then run via PBS:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell">qsub 03estimation.pbs

</code></pre></td></tr></table>
</div>
</div><h2 id="4th-differential-expression-analysis">4th, Differential expression analysis</h2>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R">
<span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">/</span><span class="n">public</span><span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">benthic</span><span class="o">/</span><span class="n">Data_benthic</span><span class="o">/</span><span class="n">practice</span><span class="o">/</span><span class="n">mRNA</span><span class="o">/</span><span class="n">phenodata.csv</span> <span class="o">$</span><span class="p">{</span><span class="n">HOME</span><span class="p">}</span><span class="o">/</span><span class="n">project</span><span class="o">/</span><span class="n">mRNA</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenodata.csv</span>

<span class="n">cd</span> <span class="o">$</span><span class="p">{</span><span class="n">HOME</span><span class="p">}</span><span class="o">/</span><span class="n">project</span><span class="o">/</span><span class="n">mRNA</span><span class="o">/</span><span class="n">output</span>

<span class="n">cat</span> <span class="o">&lt;&lt;</span><span class="n">EOF</span> <span class="o">&gt;&gt;</span><span class="n">deseq.r</span>

<span class="nf">library</span><span class="p">(</span><span class="s">&#34;DESeq2&#34;</span><span class="p">)</span>

<span class="n">path</span><span class="o">=</span><span class="nf">getwd</span><span class="p">()</span>
<span class="n">countData</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="nf">read.csv</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="s">&#34;/transcript_count_matrix.csv&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">),</span> <span class="n">row.names</span><span class="o">=</span><span class="s">&#34;transcript_id&#34;</span><span class="p">))</span>
<span class="n">colData</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="s">&#34;/phenodata.csv&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#34;,&#34;</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">all</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">colData</span><span class="p">)</span> <span class="o">%in%</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">countData</span><span class="p">))</span>
<span class="n">countData</span> <span class="o">&lt;-</span> <span class="n">countData[</span><span class="p">,</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">colData</span><span class="p">)</span><span class="n">]</span>
<span class="nf">all</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">colData</span><span class="p">)</span> <span class="o">==</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">countData</span><span class="p">))</span>

<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span> <span class="o">=</span> <span class="n">countData</span><span class="p">,</span> <span class="n">colData</span> <span class="o">=</span> <span class="n">colData</span><span class="p">,</span> <span class="n">design</span> <span class="o">=</span> <span class="o">~</span> <span class="n">group</span><span class="p">)</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>

<span class="c1"># filter count&lt;10</span>
<span class="n">dds_filter</span> <span class="o">&lt;-</span> <span class="n">dds[</span> <span class="nf">rowSums</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">))</span> <span class="o">&gt;</span> <span class="m">10</span><span class="p">,</span> <span class="n">]</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">dds_filter</span><span class="p">)</span>

<span class="c1"># plot PCA</span>
<span class="n">rld</span> <span class="o">&lt;-</span> <span class="nf">rlog</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="nf">pdf</span><span class="p">(</span><span class="s">&#34;pca_t.pdf&#34;</span><span class="p">)</span>
<span class="nf">plotPCA</span><span class="p">(</span><span class="n">rld</span><span class="p">,</span> <span class="n">intgroup</span><span class="o">=</span><span class="s">&#34;group&#34;</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1"># take group H_Red and H_YX for examples</span>
<span class="n">res</span><span class="o">&lt;-</span> <span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;group&#34;</span><span class="p">,</span><span class="s">&#34;H_Red&#34;</span><span class="p">,</span><span class="s">&#34;C_Red&#34;</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="m">0.05</span><span class="p">)</span>
<span class="n">resOrdered</span> <span class="o">&lt;-</span> <span class="n">res</span><span class="nf">[order</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="p">),</span> <span class="n">]</span>
<span class="n">diff_gene_deseq2</span> <span class="o">&lt;-</span><span class="nf">subset</span><span class="p">(</span><span class="n">resOrdered</span><span class="p">,</span><span class="n">padj</span> <span class="o">&lt;</span> <span class="m">0.05</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">log2FoldChange</span> <span class="o">&gt;</span> <span class="m">1</span> <span class="o">|</span> <span class="n">log2FoldChange</span> <span class="o">&lt;</span> <span class="m">-1</span><span class="p">))</span>
<span class="n">diff_gene_deseq2</span> <span class="o">&lt;-</span> <span class="nf">merge</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">diff_gene_deseq2</span><span class="p">),</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="n">normalize</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span> <span class="n">by</span><span class="o">=</span><span class="s">&#34;row.names&#34;</span><span class="p">,</span><span class="n">sort</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">diff_gene_deseq2</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&#34;./Red_HvC.sig.csv&#34;</span><span class="p">,</span> <span class="n">row.names</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>

<span class="n">EOF</span>

<span class="n">qsub</span> <span class="o">-</span><span class="n">I</span> <span class="o">-</span><span class="n">N</span> <span class="n">DESeq2</span> <span class="o">-</span><span class="n">l</span> <span class="n">nodes</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="n">ppn</span><span class="o">=</span><span class="m">2</span> <span class="o">-</span><span class="n">l</span> <span class="n">walltime</span><span class="o">=</span><span class="m">1000</span><span class="o">:</span><span class="m">00</span><span class="o">:</span><span class="m">00</span> <span class="o">-</span><span class="n">q</span> <span class="n">high</span> <span class="n">`Rscript deseq.R`</span>

</code></pre></td></tr></table>
</div>
</div><h2 id="5th-funtional-enrichment-analysis">5th, Funtional enrichment analysis</h2>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R">
<span class="n">cat</span> <span class="o">&lt;&lt;</span><span class="n">EOF</span> <span class="o">&gt;&gt;</span><span class="n">enrichment.R</span>

<span class="n">sig_gene</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&#34;./Red_HvC.sig.csv&#34;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>
<span class="n">universe</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&#34;/public/home/benthic/Data_benthic/practice/mRNA/Hdh_universe.txt&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;\t&#34;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">go2gene</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&#34;/public/home/benthic/Data_benthic/practice/mRNA/Hdh_go2gene.txt&#34;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#34;\t&#34;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">go2name</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&#34;/public/home/benthic/Data_benthic/practice/mRNA/Hdh_go2term.txt&#34;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#34;\t&#34;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">map2gene</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&#34;/public/home/benthic/Data_benthic/practice/mRNA/Hdh_map2gene.txt&#34;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#34;\t&#34;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">map2name</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&#34;/public/home/benthic/Data_benthic/practice/mRNA/Hdh_map2term.txt&#34;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#34;\t&#34;</span><span class="p">,</span><span class="n">header</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="s">&#34;clusterProfiler&#34;</span><span class="p">)</span>

<span class="c1"># ge enrichment</span>
<span class="n">go</span><span class="o">=</span><span class="nf">enricher</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="n">sig_gene[</span><span class="p">,</span><span class="m">1</span><span class="n">]</span><span class="p">,</span>
						<span class="n">universe</span> <span class="o">=</span> <span class="n">universe[</span><span class="p">,</span><span class="m">1</span><span class="n">]</span><span class="p">,</span>
						<span class="n">TERM2GENE</span><span class="o">=</span> <span class="n">go2gene</span><span class="p">,</span>
						<span class="n">TERM2NAME</span> <span class="o">=</span> <span class="n">go2name[</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="n">]</span><span class="p">)</span>

<span class="nf">pdf</span><span class="p">(</span><span class="s">&#34;go_enrich.pdf&#34;</span><span class="p">)</span>
<span class="nf">barplot</span><span class="p">(</span><span class="n">go</span><span class="p">,</span><span class="n">showcatergory</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
<span class="c1">#View(data.frame(go))</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1"># kegg enrichment</span>
<span class="n">kegg</span><span class="o">=</span><span class="nf">enricher</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="n">sig_gene[</span><span class="p">,</span><span class="m">1</span><span class="n">]</span><span class="p">,</span>
							<span class="n">universe</span> <span class="o">=</span> <span class="n">universe[</span><span class="p">,</span><span class="m">1</span><span class="n">]</span><span class="p">,</span>
							<span class="n">TERM2GENE</span><span class="o">=</span> <span class="n">map2gene</span><span class="p">,</span>
							<span class="n">TERM2NAME</span> <span class="o">=</span> <span class="n">map2name</span><span class="p">)</span>

<span class="nf">pdf</span><span class="p">(</span><span class="s">&#34;kegg_enrich.pdf&#34;</span><span class="p">)</span>
<span class="nf">barplot</span><span class="p">(</span><span class="n">kegg</span><span class="p">,</span><span class="n">showcatergory</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="n">EOF</span>

<span class="n">qsub</span> <span class="o">-</span><span class="n">I</span> <span class="o">-</span><span class="n">N</span> <span class="n">enrichment</span> <span class="o">-</span><span class="n">l</span> <span class="n">nodes</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="n">ppn</span><span class="o">=</span><span class="m">2</span> <span class="o">-</span><span class="n">l</span> <span class="n">walltime</span><span class="o">=</span><span class="m">1000</span><span class="o">:</span><span class="m">00</span><span class="o">:</span><span class="m">00</span> <span class="o">-</span><span class="n">q</span> <span class="n">high</span> <span class="n">`Rscript enrichment.R`</span>

</code></pre></td></tr></table>
</div>
</div><h2 id="reference">Reference</h2>
<ul>
<li><a href="https://www.nature.com/articles/nprot.2016.095" target="_blank" rel="noopener noreffer">https://www.nature.com/articles/nprot.2016.095</a></li>
<li><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" target="_blank" rel="noopener noreffer">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></li>
<li><a href="https://guangchuangyu.github.io/software/clusterProfiler/" target="_blank" rel="noopener noreffer">https://guangchuangyu.github.io/software/clusterProfiler/</a></li>
</ul>
</div>
    <div style="padding: 10px 1; margin: 10px auto; width: 100%; font-size:16px; text-align: center;">
        <button id="rewardButton" disable="enable" onclick="var qr = document.getElementById('QR'); if (qr.style.display === 'none') {qr.style.display='block';} else {qr.style.display='none'}">
            <span>Buy Me A Coffee</span></button>
        <div id="QR" style="display: none;">
            <div id="wechat" style="display: inline-block">
                <a class="fancybox" rel="group">
                    <img id="wechat_qr" src="/images/coffee.png" alt="WeChat Pay"></a>
            </div>
        </div>
    </div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>The article was updated on 2020-05-08</span>
            </div>
            <div class="post-info-license"></div>
        </div>
        <div class="post-info-line">
            <div class="post-info-md"><span>
                            <a class="link-to-markdown" href="/mrna_ref1/index.md" target="_blank">Read Markdown</a>
                        </span></div>
            <div class="post-info-share">
                <span><a href="javascript:void(0);" title="Share on Twitter" data-sharer="twitter" data-url="https://zkhuang2012.github.io/mrna_ref1/" data-title="RNA-seq Ref-align Analysis (1)" data-hashtags="RNA-seq,HISAT2,clusterProfiler"><i class="fab fa-twitter fa-fw"></i></a><a href="javascript:void(0);" title="Share on 微博" data-sharer="weibo" data-url="https://zkhuang2012.github.io/mrna_ref1/" data-title="RNA-seq Ref-align Analysis (1)"><i class="fab fa-weibo fa-fw"></i></a><a href="javascript:void(0);" title="Share on Evernote" data-sharer="evernote" data-url="https://zkhuang2012.github.io/mrna_ref1/" data-title="RNA-seq Ref-align Analysis (1)"><i class="fab fa-evernote fa-fw"></i></a></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"><i class="fas fa-tags fa-fw"></i>&nbsp;<a href="/tags/rna-seq/">RNA-seq</a>,&nbsp;<a href="/tags/hisat2/">HISAT2</a>,&nbsp;<a href="/tags/clusterprofiler/">clusterProfiler</a></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">Back</a></span>&nbsp;|&nbsp;<span><a href="/">Home</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="/about/" class="prev" rel="prev" title="About this Blog"><i class="fas fa-angle-left fa-fw"></i>About this Blog</a></div>
</div>
<div id="comments"><div id="gitalk" class="comment"></div><noscript>
                Please enable JavaScript to view the comments powered by <a href="https://github.com/gitalk/gitalk"></a>Gitalk</a>.
            </noscript></div></article></div>
            </main><footer class="footer">

        <div class="footer-container"><div class="footer-line">Powered by <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.89.4">Hugo</a> | Theme - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.6"><i class="far fa-kiss-wink-heart fa-fw"></i> LoveIt</a>
                </div><div class="footer-line"><i class="far fa-copyright fa-fw"></i><span itemprop="copyrightYear">2019 - 2024</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="/" target="_blank">zkhuang</a></span>&nbsp;|&nbsp;<span class="license"><a rel="license external nofollow noopener noreffer" href="https://creativecommons.org/licenses/by-nc/4.0/" target="_blank">CC BY-NC 4.0</a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="Back to Top">
                <i class="fas fa-arrow-up fa-fw"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="View Comments">
                <i class="fas fa-comment fa-fw"></i>
            </a>
        </div><link rel="stylesheet" href="/lib/gitalk/gitalk.min.css"><script type="text/javascript">
    window.config = {"code":{"copyTitle":"Copy to clipboard","maxShownLines":100},"comment":{"gitalk":{"admin":["zkhuang2012"],"clientID":"01f644e8e5ea430a4f00","clientSecret":"1833496f1d9d55741901203f221b9b2a749214cc","id":"2020-05-08T21:30:50+08:00","owner":"zkhuang2012","repo":"zkhuang2012.github.io","title":"RNA-seq Ref-align Analysis (1)"}},"data":{"id-1":"Huang ZK's Blog","id-2":"Huang ZK's Blog"},"headerMode":{"desktop":"fixed","mobile":"auto"},"search":{"highlightTag":"em","lunrIndexURL":"/index.json","maxResultLength":10,"noResultsFound":"No results found","snippetLength":30,"type":"lunr"},"typeit":{"cursorChar":"|","cursorSpeed":1000,"data":{"id-1":["id-1"],"id-2":["id-2"]},"duration":-1,"speed":100}};
</script><script type="text/javascript" src="https://polyfill.io/v3/polyfill.min.js?features=html5shiv%2CElement.prototype.closest%2CrequestAnimationFrame%2CCustomEvent%2CPromise%2CObject.entries%2CObject.assign%2CObject.values%2Cfetch%2CElement.prototype.after%2CArray.prototype.fill%2CIntersectionObserver%2CArray.from%2CArray.prototype.find%2CMath.sign"></script><script type="text/javascript" src="/lib/gitalk/gitalk.min.js"></script><script type="text/javascript" src="/lib/smooth-scroll/smooth-scroll.min.js"></script><script type="text/javascript" src="/lib/autocomplete/autocomplete.min.js"></script><script type="text/javascript" src="/lib/lunr/lunr.min.js"></script><script type="text/javascript" src="/lib/lazysizes/lazysizes.min.js"></script><script type="text/javascript" src="/lib/object-fit-images/ofi.min.js"></script><script type="text/javascript" src="/lib/clipboard/clipboard.min.js"></script><script type="text/javascript" src="/lib/sharer/sharer.min.js"></script><script type="text/javascript" src="/lib/typeit/typeit.min.js"></script><script type="text/javascript" src="/js/theme.min.js"></script></body>
</html>
